punchcut/src/source/glyf/deltas.rs

```diff
@@ -0,0 +1,289 @@
+        // Search for first point that has a delta.
+        while point_ix <= end_point_ix && !flags.get(point_ix)?.has_marker(PointMarker::HAS_DELTA) {
+            point_ix += 1;
+        }
+        if point_ix <= end_point_ix {
+            let first_delta_ix = point_ix;
+            let mut cur_delta_ix = point_ix;
+            point_ix += 1;
+            // Search for next point that has a delta...
+            while point_ix <= end_point_ix {
+                if flags.get(point_ix)?.has_marker(PointMarker::HAS_DELTA) {
+                    // ... and interpolate intermediate points.
+                    interpolate_range(
+                        points,
+                        cur_delta_ix + 1,
+                        point_ix - 1,
+                        cur_delta_ix,
+                        point_ix,
+                        out_points,
+                    );
+                    cur_delta_ix = point_ix;
+                }
+                point_ix += 1;
+            }
+            // If we only have a single delta, shift the contour.
+            if cur_delta_ix == first_delta_ix {
+                shift_range(
+                    points,
+                    first_point_ix,
+                    end_point_ix,
+                    cur_delta_ix,
+                    out_points,
+                )?;
+            } else {
+                // Otherwise, handle remaining points at beginning and end of contour.
+                interpolate_range(
+                    points,
+                    cur_delta_ix + 1,
+                    end_point_ix,
+                    cur_delta_ix,
+                    first_delta_ix,
+                    out_points,
+                )?;
+                if first_delta_ix > 0 {
+                    interpolate_range(
+                        points,
+                        first_point_ix,
+                        first_delta_ix - 1,
+                        cur_delta_ix,
+                        first_delta_ix,
+                        out_points,
+                    )?;
+                }
+            }
+        }
+    }
+    Some(())
+}
+
+/// Shift the coordinates of all points between `start` and `end` inclusive
+/// using the difference given by the point at `ref_`.
+///
+/// Modeled after the FreeType implementation: <https://github.com/freetype/freetype/blob/bbfcd79eacb4985d4b68783565f4b494aa64516b/src/truetype/ttgxvar.c#L3776>
+fn shift_range(
+    points: &[Point<i32>],
+    start: usize,
+    end: usize,
+    ref_: usize,
+    out_points: &mut [Point<Fixed>],
+) -> Option<()> {
+    let ref_in = points.get(ref_)?.map(Fixed::from_i32);
+    let ref_out = out_points.get(ref_)?;
+    let delta = *ref_out - ref_in;
+    if delta.x == Fixed::ZERO && delta.y == Fixed::ZERO {
+        return Some(());
+    }
+    for out_point in out_points.get_mut(start..ref_)? {
+        *out_point += delta;
+    }
+    for out_point in out_points.get_mut(ref_ + 1..=end)? {
+        *out_point += delta;
+    }
```

Comment:
the two loops is weird, you specifically don't want to shift ref by itself?

Reply:
Yes, we only call shift on a contour that contains a single point with an applied delta. In this case, we apply that same delta to the remaining points. The loops could be combined with a `split_at_mut` and a `chain`, but I think that is less clear than the double loop. I'll add a comment to explain what's going on.